contracts/verifiers/Verifier.sol

```diff
@@ -0,0 +1,976 @@
+error AuthMethodAlreadyExists(string authMethod);
+error GroupIdNotFound(uint256 groupId);
+error GroupIdAlreadyExists(uint256 groupId);
+error GroupMustHaveAtLeastTwoRequests(uint256 groupID);
+error LinkIDNotTheSameForGroupedRequests();
+error MetadataNotSupportedYet();
+error MultiRequestIdAlreadyExists(uint256 multiRequestId);
+error MultiRequestIdNotFound(uint256 multiRequestId);
+error NullifierSessionIDAlreadyExists(uint256 nullifierSessionID);
+error ResponseFieldAlreadyExists(string responseFieldName);
+error RequestIdAlreadyExists(uint256 requestId);
+error RequestIdNotFound(uint256 requestId);
+error RequestIdNotValid();
+error RequestIdUsesReservedBytes();
+error RequestIdTypeNotValid();
+error RequestShouldNotHaveAGroup(uint256 requestId);
+error UserIDMismatch(uint256 userIDFromAuth, uint256 userIDFromResponse);
+error UserIDNotFound(uint256 userID);
+error UserIDNotLinkedToAddress(uint256 userID, address userAddress);
+error UserNotAuthenticated();
+error ValidatorNotWhitelisted(address validator);
+error VerifierIDIsNotValid(uint256 requestVerifierID, uint256 expectedVerifierID);
+error ChallengeIsInvalid();
+
+abstract contract Verifier is IVerifier, ContextUpgradeable {
+    /// @dev Key to retrieve the linkID from the proof storage
+    string private constant LINKED_PROOF_KEY = "linkID";
+    // keccak256(abi.encodePacked("authV2"))
+    bytes32 private constant AUTHV2_NAME =
+        0x380ee2d21c7a4607d113dad9e76a0bc90f5325a136d5f0e14b6ccf849d948e25;
+    // keccak256(abi.encodePacked("challenge"))
+    bytes32 private constant CHALLENGE_NAME =
+        0x62357b294ca756256b576c5da68950c49d0d1823063551ffdcc1dad9d65a07a6;
+
+    struct AuthMethodData {
+        IAuthValidator validator;
+        bytes params;
+        bool isActive;
+    }
+
+    /// @custom:storage-location erc7201:iden3.storage.Verifier
+    struct VerifierStorage {
+        // Information about requests
+        // solhint-disable-next-line
+        mapping(uint256 requestId => mapping(address sender => Proof)) _proofs;
+        mapping(uint256 requestId => IVerifier.RequestData) _requests;
+        uint256[] _requestIds;
+        IState _state;
+        mapping(uint256 groupId => uint256[] requestIds) _groupedRequests;
+        uint256[] _groupIds;
+        // Information about multiRequests
+        mapping(uint256 multiRequestId => IVerifier.MultiRequest) _multiRequests;
+        uint256[] _multiRequestIds;
+        // Information about auth methods and validators
+        string[] _authMethodsNames;
+        mapping(string authMethod => AuthMethodData) _authMethods;
+        mapping(uint256 nullifierSessionID => uint256 requestId) _nullifierSessionIDs;
+        // verifierID to check in requests
+        uint256 _verifierID;
+    }
+
+    /**
+     * @dev Struct to store proof and associated data
+     */
+    struct Proof {
+        bool isVerified;
+        mapping(string key => uint256 inputValue) storageFields;
```

Comment:
Rename to responseFields in 3 places
And this ordering may be better
   ```
 struct Proof {
        bool isVerified;
        mapping(string key => uint256 inputValue) storageFields;
        string[] storageFieldNames;
        // introduce artificial shift + 1 to avoid 0 index
        mapping(string key => uint256 keyIndex) storageFieldIndexes;
        string validatorVersion;
        uint256 blockTimestamp;
        uint256[44] __gap;
    }
```

Reply:
Done